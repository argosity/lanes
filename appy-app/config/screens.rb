
Lanes::Screen.define "base" do | screen |
    screen.description = "Base Screen for Appy App"
    screen.icon        = ""
    screen.group_id    = ""
    screen.model_class = ""
    screen.view_class  = "AppyApp.Screens.Base"
end
