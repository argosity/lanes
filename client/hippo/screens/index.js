import { observable, computed } from 'mobx';
import { map, filter, intersection } from 'lodash';
import Group from './group';
import Config from '../config';

const Screens = observable({

    all: observable.map(),

    get activeGroups() {
        const gids = Group.enabled_group_ids || map(Group.all, 'id');
        const groups = map(gids, gid => Group.forId(gid));
        return filter(groups, group => intersection(group.screens, this.active).length);
    },

    @computed get active() {
        return map(Config.screen_ids, id => this.all.get(id));
    },

    reset() {
        Config.screen_ids.clear();
    },

});

export default Screens;
