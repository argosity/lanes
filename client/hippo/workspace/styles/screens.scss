.screen {
    padding: 5px;
    width: 100%;

    > .first-run {
        font-size: 1.3rem;
        li + li {
            margin-top: 1rem;
        }
        .help {
            margin-top: 2rem;
            p { text-align: right; }
        }
    }
}
