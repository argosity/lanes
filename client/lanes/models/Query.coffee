##= require_self
##= require ./query/Result
##= require ./query/ArrayResult
##= require ./query/CollectionResult

class Field extends Lanes.Models.Base
    registerForPubSub: false

    constructor: (attributes) ->
        super( _.defaults( attributes, {
            title: _.titleize(_.humanize(attributes.id))
        }))

    session:
        id:         type: 'string'
        title:      type: 'string'
        selected:   type: 'boolean'
        visible:    type: 'boolean', default: true
        query:      type: 'boolean', default: true
        editable:   type: 'boolean', default: true
        sortable:   type: 'boolean', default: true
        format:     type: 'function'
        flex:       type: 'number',  default: 1
        fixedWidth: type: 'number'
        textAlign:  type: 'string', default: 'left'
        sortBy:     type: 'any'
        component:  type: 'function'
        onColumnClick: type: 'function'

    derived:
        default_value: deps: ['id'], fn: ->
            switch @model_field.type
                when 'integer' then 0
                else ''
        model_field:
            deps: ['id'], fn: ->
                this.collection.query.model::_definition[@id]
        type:
            deps: ['model_field'], fn: ->
                type = @model_field?.type || 'string'
                if type == "code" then "string" else type
        fetchIndex:
            fn: ->
                fetchIndex = 0
                for field, index in this.collection.models
                    return fetchIndex if this == field
                    fetchIndex += 1 if field.query

    validValue: (value) ->
        if this.type == 'n'
            ! _.isNaN( parseFloat(value) )
        else
            value


class AvailableFields extends Lanes.Models.Collection

    model: Field

    constructor: (models, options) ->
        @query = options.query
        this.on('change', (changing) ->
            return unless changing.selected
            for model in @models
                model.selected = false unless model == changing
        )
        super
        @visible = @subcollection(where: {visible: true})


class Operator extends Lanes.Models.Base
    registerForPubSub: false

    session:
        id:       'string'
        name:     'string'
        types:    'array'
        field:    'object'
        selected: 'boolean'

    validForField: (field) ->
        _.isEmpty(@types) or _.includes(this.types, field.type)


class Operators extends Lanes.Models.Collection
    model: Operator

    constructor: (models, attrs) ->
        super
        this.add([
            { id: 'like', name: 'Starts With', types: Lanes.Models.Query.LIKE_QUERY_TYPES }
            { id: 'eq',   name: 'Equals' }
            { id: 'lt',   name: 'Less Than', types: Lanes.Models.Query.LESS_THAN_QUERY_TYPES }
            { id: 'gt',   name: 'More Than', types: Lanes.Models.Query.LESS_THAN_QUERY_TYPES }
        ])
        this.on('change', (changing) ->
            return unless changing.selected
            for model in @models
                model.selected = false unless model == changing
        )
        @setField(attrs.parent.query.initialField)

    setField: (field) ->
        @field = field
        @valid = @subcollection(filter: (op) => op.validForField(@field) )
        selected = this.findWhere(selected: true)
        unless selected && selected.validForField(field)
            @valid.at(0)?.selected = true


class Clause extends Lanes.Models.Base
    registerForPubSub: false

    session:
        value     : { type: 'string', default: '' }
        field     : 'state'
        operator  : 'state'

    associations:
        operators : { collection: Operators }
        fields:     { collection: AvailableFields }

    derived:
        description:
            deps: ['field', 'operator']
            fn: -> "#{@field.title} #{@operator?.id}"

        isValid:
            deps: ['field', 'operator', 'value']
            fn: -> this.field.validValue(@value)

        query:
            fn: -> @collection.query

    constructor: (options) ->
        super
        this.fields.reset(options.available_fields.models)
        @fields.on('change:selected',    this.setField,    this)
        @operators.on('change:selected', this.setOperator, this)
        @operator = @operators.findWhere(selected: true)
        if (field = @fields.findWhere(visible: true))
            field.selected = true
            @setField(field)

    setFromView: (type, val) ->
        if type == "fields" || type == "operators"
            this[type].get(val).selected = true
        else
            super

    setField: (field) ->
        return unless field.selected
        @operators.setField( field )
        @field = field

    setOperator: (operator) ->
        return unless operator.selected
        this.operator = operator

    remove: ->
        @collection.remove(this)

    toParam: ->
        param = {}
        op = this.operator.id
        value = this.get('value')
        value += '%' if 'like' == op
        value = parseFloat(value) if @field.type == "n"
        param[ @field.id ] = if 'eq' == op then value else { op: op, value: value }
        param


class Clauses extends Lanes.Models.Collection

    model: Clause

    initialize:(models, options) ->
        super
        @query = options.query
        @fields = options.query.fields

    session:
        field: 'state'

# needs to inherit from Base so network events will be listened to
class Lanes.Models.Query extends Lanes.Models.Base
    registerForPubSub: false

    @LIKE_QUERY_TYPES: ['string', 'code', 'visible_id']
    @LESS_THAN_QUERY_TYPES: ['integer', 'bigdec', 'number']
    @GREATER_THAN_QUERY_TYPES: ['integer', 'bigdec', 'number']

    @mergedSyncOptions: (args...) ->
        _.merge {}, args..., (a, b) ->
            return a.concat(b) if _.isArray(a)

    session:
        src:     'any'
        initialField: 'state'
        idIndex: 'number'
        initialFieldIndex: 'number'
        pageSize: { type: 'number', default: 100 }
        syncOptions: 'any'
        autoRetrieve: {type:'boolean', default: false, required: true}
        title: { type: 'string', default: 'Find Record' }
        sortField: 'object'
        defaultSort: 'any'
        changeCount: {type: 'integer', default: 0}
        sortAscending: ['boolean', true, true]

    associations:
        fields:  {collection: AvailableFields, inverse: 'query'}
        clauses: {collection: Clauses, inverse: 'query'}

    derived:
        isCollection:
            deps: ['src'], fn: -> Lanes.u.isCollection(@src)

        model:
            deps: ['src'], fn: ->
                return null unless @src
                if @isCollection then @src.model else @src

        idAttribute:
            deps: ['model'], fn: -> @model::idAttribute

        results:
            deps: ['src'], fn: ->
                if @isCollection
                    new Lanes.Models.Query.CollectionResult(this)
                else
                    new Lanes.Models.Query.ArrayResult(this, pageSize: @pageSize)

    events:
        'change:sortField':     '_updateSort'
        'change:sortAscending': '_updateSort'

    constructor: (options = {}) ->
        super

        @fields.reset()
        for col, i in options.fields
            rec = if _.isObject(col) then col else { id: col }
            @fields.add rec
            @idIndex = i if rec.id == @idAttribute
        unless @idIndex?
            @idIndex = @fields.length
            @fields.add(id: @idAttribute)
        # @clauses = new Clauses([], query: this )
        this.listenTo(@clauses, 'change remove reset', =>
            @results.reset()
            @results.ensureLoaded() if this.autoRetrieve
        )
        this.on('change:src change:results', ->
            this.trigger('load')
        )
        this.on('load', ->
            @changeCount += 1
        )

        if @initialFieldIndex
            @initialField = this.fields.visible.at(@initialFieldIndex)

        @initialField ||= this.fields.visible.findWhere(id: "code") ||
            this.fields.visible.findWhere(id: "visible_id") ||
            this.fields.visible.first()
        @reset(true)
        this

    clonedAttributes: ->
        attrs = @getAttributes(session: true, derived: false)
        attrs.fields = this.fields.serialize({session: true})
        attrs

    reset: (silent = false) ->
        unless @defaultSort is false
            sort = @defaultSort or @fields.findWhere(visible: true).id
            @setSortField( @fields.findWhere(id: sort), sortAscending: @sortAscending, silent: true )

        @clauses.reset([
            {query: this, available_fields: @fields, field: @initialField}
        ], {silent})

    setSortField: (field, options = {silent: false}) ->
        options.sortAscending ?= (if @sortField is field then !@sortAscending else true)
        @set({
            sortAscending: options.sortAscending
            sortField: field
        }, options)

    ensureLoaded: -> @results.ensureLoaded()

    isValid: ->
        ! @clauses.findWhere( isValid: false )

    loadModel: (model, options = {}) ->
        _.extend(options, _.result(this, 'syncOptions'), force: true)
        model.withAssociations(options.include || [], options)

    loadSingle: (code, options = {}) ->
        options.query = {}
        options.query[ @initialField.id ] = code
        _.extend(options, _.result(this, 'syncOptions'))
        @trigger('request')
        @src.fetch(options).then (model) =>
            @trigger('load')
            model

    defaultField: ->
        @fields.findWhere( field: @initialField )

    addNewClause: ->
        @clauses.add(query: this, available_fields: @fields, field: @initialField)

    _updateSort: _.debounce ->
        @results._updateSort()
