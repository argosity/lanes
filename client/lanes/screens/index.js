import { observable, autorun } from 'mobx';
import { map, uniq } from 'lodash';
import Group from './group';
import Config from '../config';
import Sync from '../models/sync';

import user from '../user';

const Screens = observable({

    all: observable.map(),

    active: observable.array(),

    groups: observable.map(),

    get activeGroups() {
        return uniq(map(this.active, s => Group.forId(s.group_id)));
    },

    refresh() {
        Sync.perform(`${Config.api_path}/lanes/screens`).then(({ data: { screens: screenIds } }) => {
            this.active.replace(map(screenIds, id => this.all.get(id)));
        });
    },

});
export default Screens;

let previousLoggedIn = null;
autorun(() => {
    if (user.isLoggedIn !== previousLoggedIn){
        previousLoggedIn = user.isLoggedIn;
        Screens.refresh();
    }
});
