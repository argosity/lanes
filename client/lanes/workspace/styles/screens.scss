.screen {
    padding: 5px;
    //  overflow: hidden;
    width: 100%;
    .toolbar {
        margin-top: 10px;
    }

    .toolbar{
        @extend .navbar;
        @extend .navbar-default;

        .btn {
            @extend .btn-default;
            @extend .navbar-btn;
        }

        .arrow {
            border-style: solid;

            border-left-color: transparent;
            border-right-color: transparent;

            border-bottom-color: $popover-title-bg;
            border-top-color: $popover-title-bg !important; 

            border-width: 0 9px 9px;
            height: 0;
            left: 7px;
            margin-top: 10px;
            opacity: 1;
            position: absolute;
            top: -18px;
            transition: all 0.25s ease 0s;
            width: 0;
            z-index: 10;
        }
        .navbar-left {
            float: left;
        }
        .navbar-right{
            float: right;
            .arrow {
                left: unset;
                right: 10px;
            }
        }

        .dropdown-toggle {
            .badge{
                padding: 2px 2px 0;
                position: relative;
                font-size: 70%;
                top: -6px;
                right: 5px;
            }
            position: relative;
            padding-right: 2px;
            i {
                /* position: absolute; */
                /* left: 0px; */
            }
        }

    }

}
