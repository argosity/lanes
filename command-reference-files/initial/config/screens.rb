# An example of how to define a screen group
# Hippo::Screen.define_group 'appy-app-screens' do | group |
#     group.title       = "AppyApp Screens"
#     group.description = "Screens relating to AppyApp"
#     group.icon        = "heart"
# end

Hippo::Screen.for_extension 'appy-app' do | screens |

end
