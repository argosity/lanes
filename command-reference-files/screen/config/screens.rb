# An example of how to define a screen group
# Hippo::Screen.define_group 'appy-app-screens' do | group |
#     group.title       = "AppyApp Screens"
#     group.description = "Screens relating to AppyApp"
#     group.icon        = "heart"
# end

Hippo::Screen.for_extension 'appy-app' do | screens |
    screens.define "ready-set-go" do | screen |
        screen.title       = "Ready Set Go"
        screen.description = ""
        screen.icon        = ""
        screen.group_id    = ""
        screen.model_class = ""
        screen.view_class  = "ReadySetGo"
    end

end
