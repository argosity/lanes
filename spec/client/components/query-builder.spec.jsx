import React from 'react';

import QueryBuilder from 'lanes/components/query-builder';

import { Box } from '../test-models';
import { Snapshot } from 'lanes/testing/screens';
import Query from 'lanes/models/query';

jest.useFakeTimers();

describe("Query Builder Component", () => {

    let query;

    beforeEach(() => {
        query = new Query({
            src: Box,
            fields: [
                { id: 'id', visible: false, queryable: false },
                { id: 'computed', title: 'IsAdmin?', loadable: false },
                { id: 'label' },
                { id: 'width' },
                { id: 'height' },
                { id: 'depth', visible: false },
            ],
        });
    });

    it('renders and matches snapshot', () => {
        expect(Snapshot(<QueryBuilder query={query} />)).toMatchSnapshot();
    });

    it('autoloads query when changed', () => {
        query.fetch = jest.fn();
        expect(query.autoFetch).toBe(false);
        const builder = mount(<QueryBuilder query={query} autoFetch />);
        expect(query.autoFetch).toBe(true);
        expect(query.fetch).toHaveBeenCalled();
        query.clauses[0].value = 'test';
        jest.runAllTimers();
        expect(query.fetch).toHaveBeenCalledTimes(2);
        builder.unmount();
        expect(query.autoFetch).toBe(false);
    });
});
