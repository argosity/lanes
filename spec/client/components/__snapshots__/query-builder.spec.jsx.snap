// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query Builder Component renders and matches snapshot 1`] = `
.c3 {
  color: #865CD6;
  text-decoration: none;
  cursor: pointer;
  outline: none;
}

.c3:hover {
  text-decoration: underline;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  max-width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c5 {
  padding: 11px;
  border: 1px solid rgba(0,0,0,0.15);
  border-radius: 4px;
  outline: none;
  background-color: transparent;
  color: inherit;
  font: inherit;
  margin: 0;
}

.c5::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c5::-moz-placeholder {
  color: #AAAAAA;
}

.c5:-ms-input-placeholder {
  color: #AAAAAA;
}

.c5::-moz-focus-inner {
  border: none;
  outline: none;
}

.c5:focus {
  border-color: #00CCEB;
  box-shadow: 0 0 2px 2px #00CCEB;
}

.c5:focus > circle,
.c5:focus > ellipse,
.c5:focus > line,
.c5:focus > path,
.c5:focus > polygon,
.c5:focus > polyline,
.c5:focus > rect {
  outline: #00CCEB solid 2px;
}

.c0 {
  font-family: 'Work Sans',Arial,sans-serif;
  font-size: 1em;
  line-height: 1.5;
  color: #333333;
  background-color: #FFFFFF;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c0 * {
  box-sizing: inherit;
}

.c2 {
  cursor: pointer;
  outline: none;
  font: inherit;
  text-decoration: none;
  font: inherit;
  margin: 0;
  background-color: transparent;
  overflow: visible;
  text-transform: none;
  color: inherit;
  border: none;
  padding: 0;
  width: 100%;
  max-width: none;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c6 {
  cursor: pointer;
  outline: none;
  font: inherit;
  text-decoration: none;
  font: inherit;
  margin: 0;
  background-color: transparent;
  overflow: visible;
  text-transform: none;
  color: inherit;
  border: none;
  padding: 0;
  padding: 12px;
}

.c4 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  fill: #666666;
  stroke: #666666;
}

.c4 g {
  fill: inherit;
  stroke: inherit;
}

.c4 *:not([stroke])[fill="none"] {
  stroke-width: 0;
}

.c4 *[stroke*="#"],
.c4 *[STROKE*="#"] {
  stroke: inherit;
  fill: none;
}

.c4 *[fill-rule],
.c4 *[FILL-RULE],
.c4 *[fill*="#"],
.c4 *[FILL*="#"] {
  fill: inherit;
  stroke: none;
}

.c7 {
  display: inline-block;
}

.c7:first-child:not(:last-child) {
  margin-right: 12px;
}

.c7 > * {
  vertical-align: bottom;
}

@media only screen and (min-width:700px) {
  .c2 {
    -webkit-transition: 0.1s ease-in-out;
    transition: 0.1s ease-in-out;
  }
}

@media only screen and (min-width:700px) {
  .c6 {
    -webkit-transition: 0.1s ease-in-out;
    transition: 0.1s ease-in-out;
  }
}

<div
  className="c0"
>
  <div
    className="query-builder"
  >
    <div
      aria-label={undefined}
      className="clause c1"
      direction="row"
    >
      <button
        aria-label="Open Drop"
        className="c2"
        disabled={false}
        fill={true}
        href={undefined}
        icon={undefined}
        id={undefined}
        label={undefined}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        open={false}
        size="large"
        tabIndex="0"
        type="button"
      >
        <a
          aria-label={undefined}
          className="c3"
          disabled={undefined}
          href={undefined}
          icon={undefined}
          label={undefined}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
        >
          <svg
            aria-label="Down"
            className="c4"
            height="24px"
            role="img"
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
          >
            <polyline
              fill="none"
              points="7.086 3.174 17.086 13.174 7.086 23.174"
              stroke="#000"
              strokeWidth="2"
              transform="scale(1 -1) rotate(-89 -1.32 0)"
            />
          </svg>
           
          IsAdmin?
           
          Starts With
        </a>
      </button>
      <div
        className=""
      >
        <input
          autoComplete="off"
          autoFocus={true}
          className="c5"
          defaultValue={undefined}
          id={undefined}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          style={
            Object {
              "flex": 1,
            }
          }
          value=""
        />
      </div>
      <button
        aria-label={undefined}
        className="c6"
        disabled={false}
        href={undefined}
        icon={<Refresh />}
        label={undefined}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          aria-hidden={true}
          className="c7"
        >
          <svg
            aria-label="Refresh"
            className="c4"
            height="24px"
            role="img"
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
          >
            <path
              d="M20,8 C18.5974037,5.04031171 15.536972,3 12,3 C7.02943725,3 3,7.02943725 3,12 C3,16.9705627 7.02943725,21 12,21 L12,21 C16.9705627,21 21,16.9705627 21,12 M21,3 L21,9 L15,9"
              fill="none"
              stroke="#000"
              strokeWidth="2"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;
